Optionally hide files and folders starting with dot

mihnita opened the pull request:
## Summary

Because in the UNIX / Linux / macOS world the files and folders starting with a '.' are not shown, this convention "leaked" into Windows as well.

These files "get in the way" the same way hidden / system files do.

It would be really nice to be able to easily hide / unhide these files / folders, the same way we can do with the ones with hidden / system attributes.

## References

Issue #995

## Checklist
- [x] I have followed the [contributing guidelines]
- [ ] I have discussed this with project maintainers:<br/>
If not checked, I accept that this work might be rejected in favor of a different grand plan.<br/>

I did not discuss it, but the change is small enough that if it is rejected I will not mind.
(I might still mind the lack of that feature though ;-)

## Details

The change is quite trivial.

In most places related to hiding/showing hidden and system files I added similar code to hide/show files starting with a dot.

## Note

I was not sure how to add the extra section in the help for other languages, so I only updated `FarEng.hlf.m4`

MKadaner reviewed (commented):
Thanks.

> I was not sure how to add the extra section in the help for other languages, so I only updated `FarEng.hlf.m4`

Simply copy-paste the English paragraphs to the corresponding places in other `.hlf.m4` files. Please make sure to keep the exact paragraph structure (number of paragraphs, indentations, empty lines after).

Disclaimer: I am not a maintainer.

alabuzhev commented:
Hi @mihnita,

Thanks for looking into this.

> this convention "leaked" into Windows as well

It did and it is so frustrating.
The whole idea of dotfiles is so fundamentally broken.

Just a few examples.
No one probably wants or needs to see the .git directory, because it's just some boring metadata, managed by the system.
But at the same time .gitignore and .gitattributes are very real files that are supposed to be in the index and properly managed by you to reflect your project needs.
And on top of that the bloody Github requires you to keep your workflows and issue & pull request templates in... the .github directory. 🤦 
Why "the stuff I don't care about" and "the stuff I have to care about" are handled in the same way?
This does not make any sense.
And in a typical home directory on a typical linux machine it's exactly the same situation, but on the scale of full blown insanity: manual configs, generated configs, histories, caches, some binary rubbish, whatever - a legion of them.

And the worst thing is that the file name is, obviously, part of the contract (unlike the file attributes), so you generally cannot unhide the stuff you do care about and keep the rubbish hidden (like in Windows).
This does not make any sense.
It's as if one approach was deliberately designed, and the other happened to evolve from a bug in ls. Oh wai...

 
But I digress.

Yes, we _probably_ should support it.
The whole concept is cursed, but, as you mentioned, it's leaking all over the place and with the advent of WSL it will only get worse with time.
Even the damn Visual Studio now creates .vs directories for some unholy reasons.
@microsoft, watch out, there's an impostor in VS team! 😄 

However, I do not think that we need to do it separately with all the glory like a config option, a hotkey, an API exposure etc.
If such files, or at least a significant portion of them, are conceptually hidden / system, albeit bastardized, **why don't we just bundle them with real hidden / system files?**
As you said, they get in the way the same way hidden / system files do.
Especially since quite a few apps ported to Windows use both the dot and the hidden attribute simultaneously, so you'll probably have to press Ctrl+D Ctrl+H together more often than not anyway just to make sure.

> I might still mind the lack of that feature though ;-)

Do you know that we have panel filters (Ctrl+I), which are basically Ctrl+H on steroids, and, technically, this particular feature (and much more, like actually separating the wheat from the tares and .git from .gitattributes) has been there all along?

> I was not sure how to add the extra section in the help for other languages

We have a check during the build that the basic structure and the number of paragraphs are the same, to stop them from diverging too much.
If you need to update the help, as @MKadaner mentioned, just copy the English blocks to all the languages you're not familiar with.

HamRusTal commented on far/FarEng.hlf.m4:
```diff
@@ -407,6 +407,7 @@ $ #Panel control commands#
  Set alternative full view mode                          #LeftCtrl+0#
 
  Toggle hidden and system files displaying                   #Ctrl+H#
+ Toggle dot files and folders displaying                     #Ctrl+D#
```
Like it or not but `Ctrl+D` in panels is utilized in the command line (if not empty) already: it's equivalent to `Right`. So the suggested change will be breaking for some users (not me personally, I have a macro for `Ctrl+D` to perform another action).

HamRusTal replied:
> Убрать/показать файлы

I suggest «Скрыть» instead of «Убрать» as that will be both more technical and less ambiguous.

mihnita replied:
> Like it or not but Ctrl+D in panels is utilized in the command line (if not empty) already

ACK.
I will look for something else.
Or, if you all agree, I can go with the latest 2 suggestions: treating dot files as hidden, and only use one hotkey for both.

MKadaner replied:
If you go with the new option, please use "Скрыть/показать" everywhere. There is no good translation for "Toggle" in Russian.

mihnita replied:
Like this?
```
 Скрыть/показать файлы с атрибутом "Скрытый" и "Системный"   #Ctrl+H#
 Скрыть/показать отображение длинных/коротких имён файлов    #Ctrl+N#
```

Note:
- the "new" option (with #Ctrl+D#) is gone
- I don't know Russian. I can read, and I can understand maybe 10% :-) But no

alabuzhev replied:
Guys, please don't clutter the PR with translation peculiarities, especially since the author does not know the language and needs to wait for your input every time.
Please feel free to raise a following PR with whatever wording you prefer.

mihnita commented on far/FarEng.hlf.m4:
```diff
@@ -2504,6 +2505,9 @@ $ #Settings dialog: panel#
 
+ #Show dot files#
+ Enables displaying files and folders that start with a dot. This option can also be switched by #Ctrl+D#.
+
```
Done

mihnita commented:
Thank you all for the feedback.

I've changed the implementation to use `Panel.TreatDotFilesAsHidden`
I merged all the previous commits to hide all the back-and-forth, as this last change reverts several files (all the Plugin* stuff).

But since this required new wording I will need (again) Russian translations, sorry.

And because the previous string also changed, the Russian (and others :-( ) might also have to change.
Not critical (does not fundamentally change anything), but nice to update.

```diff
MConfigHidden
"Показывать скр&ытые и системные файлы"
"-Show &hidden and system files"
"+Show &hidden and system files and folders"
...
```
Should I add `upd:` in front of the other languages (for `MConfigHidden`)?

And is there an equivalent of `upd:` for the help files?

Thank you,
Mihai

yegor-mialyk commented:
> Do you know that we have panel filters (Ctrl+I)

Oh, I forgot about filters. Technically `Show/Hide hidden files` should have been migrated to filters + key macro :)

MKadaner commented:
If you add new `Panel.TreatDotFilesAsHidden` `far:Config`, please add help article (bunch of them at the bottom of the `.hlf.m4` file. When you have English version, I'll translate it to Russian. And you do not need any other language strings, right?

alabuzhev commented:
> But because you still need to see files like .gitconfig, .gitignore, .github, and so on, it is a lot nicer to have a hotkey to easily turn things on / off.

Do you know that we have macros and you can have a hotkey to basically anything, including switching certain filters on and off?

yegor-mialyk commented:
> I would personally prefer a setting (e.g. `Panel.TreatDotFilesAsHidden`) in `far:config`, and I’d like to keep using `Ctrl+H` as before.

Just to make it clear I meant "As a user I would prefer a setting..." above. However, my hidden files are always "tuned on" and `Ctrl+H` is remapped to a different action :)

MKadaner commented on far/farlang.templ.m4:
```diff
@@ -2607,7 +2607,7 @@ l:
 
 MConfigHidden
 "Показывать скр&ытые и системные файлы"
```
```suggestion
"Показывать скр&ытые и системные файлы и папки"
```

mihnita commented:
New commit, with all the changes requested.
So it should be ready for review.
Thank you all,
Mihai

mihnita commented:
> And if you do it, please add a help article similar to `@Panel.CtrlFRule` one.

Done.

mihnita commented on far/FarEng.hlf.m4:
```diff
@@ -6874,6 +6874,23 @@ deleted, or renamed.
+ True  - Dot files and folders will be hidden / shown together
+with the Hidden and System files.
+
+ Default value: True (hide / show the dot files in sync with the
```
Done

mihnita commented on far/FarEng.hlf.m4:
```diff
@@ -6874,6 +6874,23 @@ deleted, or renamed.
+option in the ~panel settings~@PanelSettings@ dialog.
+
+ False - ^<wrap>Dot files and folders will always be shown.
+ True  - Dot files and folders will be hidden / shown together
```
> Also, it's a very minor thing but I'd suggest hidden/shown without spaces (in all language versions of course).

Done

---

> Shouldn't we enclose “Dot files” in (curly) quotes? This is not an official term after all, and it can even be confused with files with the .dot extension…

I tried, but it was weird, because the text was - Dot files and folders...
So technically I would have had to quote  not just
` False - "Dot files and folders" will be hidden/shown ...`
not just
` False - "Dot files" and folders will be hidden/shown ...`

So I did two things:
1. Changed the first paragraph to also kind of "define" what dotfiles are
- This Boolean parameter controls the filtering of files or folders
starting with a dot (“dotfiles”).

2. After searching online, there are many-many mentions of "dotfiles", without space
Many of the first hits also contain a definition of what those are.
So I change the term to "dotfiles" (without spaces), as that seems a lot more established.

WDYT?

---

Thank you,
M

HamRusTal commented on far/farlang.templ.m4:
```diff
@@ -1,4 +1,4 @@
-#
+﻿#
```
Did something happen to the BOM here, did it dis-/appear? Won't it confuse the build pipeline tools?

HamRusTal replied:
IMHO if it's UTF-8 and BOM does not break the build, we should leave it.

MKadaner commented on far/config.cpp:
```diff
@@ -1983,6 +1983,7 @@ void Options::InitConfigsData()
 		{FSSF_PRIVATE,           NKeyPanel,                  L"ShellRightLeftArrowsRule"sv,      ShellRightLeftArrowsRule, false},
 		{FSSF_PANEL,             NKeyPanel,                  L"ShowBytes"sv,                     ShowBytes, false},
 		{FSSF_PANEL,             NKeyPanel,                  L"ShowHidden"sv,                    ShowHidden, true},
+		{FSSF_PANEL,             NKeyPanel,                  L"TreatDotFilesAsHidden"sv,         TreatDotFilesAsHidden, true},
```
Is it a good idea to change the default, out of the box behavior? Even though I think I may use this feature in the future, as a Far user, I would be very unhappy if my `.gitignore` files disappear.

MKadaner replied:
When I wear my developer's hat, I tend to agree with you. Moreover, because I am doing such-and-such feature for my own convenience, I also want to flip the default behavior, so that it is on for me out of the box. I am still not sure how I would take this behavior change as a user. Probably would be pissed off. And, since, as you aptly pointed out, the `far:config`-only option is undiscoverable, I would not even know how to fix the damn thing. 🙂

For the discoverability, as far as I have observed here, the things happen like this. Somebody comes with a question or suggestion, "I want to have this or that." And more often than not, somebody replies, "Please use this existing `far:config` (or another 'hidden') feature." Not very scalable but somehow works.

To be serious, un-discoverability of `far:config`-only options has bothered me for long time. Maybe we should add a directory of all articles covering such options to the Help? It probably can be composed automatically and linked from the main Help contents page.

@alabuzhev, what do you think? Yes, I am aware that this list can be seen on the `Help file index` page, as well as that the Help can be searched for `far:config` to get more or less same result. I would argue, however, that the discoverability of either of these ways is also less than great.

mihnita replied:
Done (changed to `false`)

MKadaner replied:
Thank you.

MKadaner commented on far/FarEng.hlf.m4:
```diff
@@ -6874,6 +6874,22 @@ deleted, or renamed.
+$ #far:config Panel.TreatDotFilesAsHidden#
+ This Boolean parameter controls the filtering of files or folders
+starting with a dot (“dotfiles”).
+ ^<wrap>It works in combination with the #Show hidden and system files and folders#
```
It is better to use exact phrase as it appears in the dialog (and in the help above): `#Show hidden and system files#`. The "files and folder" wording is more precise, and we should probably adopt it, but we should be consistent.

MKadaner replied:
Thanks!

mihnita commented on far/FarEng.hlf.m4:
```diff
@@ -6874,6 +6874,22 @@ deleted, or renamed.
+$ #far:config Panel.TreatDotFilesAsHidden#
+ This Boolean parameter controls the filtering of files or folders
+starting with a dot (“dotfiles”).
+ ^<wrap>It works in combination with the #Show hidden and system files and folders#
```
Done

MKadaner replied:
Thank you.

MKadaner commented on far/aa:
Oops!

MKadaner replied:
Thanks.

mihnita reviewed (commented):
Implemented everything.
Defaulting to false, fixed help.

---

Let me know if you want me to squash the commits into one, for cleaner history.
Or you can squash?
I don't know the customs in this repo :-)

Thank you all,
Mihai

mihnita commented on far/FarEng.hlf.m4:
```diff
@@ -6874,6 +6874,22 @@ deleted, or renamed.
+ False - ^<wrap>Dotfiles will always be shown.
+ True  - Dotfiles will be hidden/shown together with the Hidden and System files.
+
+ Default value: True (hide/show the dotfiles in sync with the
+Hidden and System files).
```
Darn :-)
Thanks, will do in a second!

MKadaner commented:
> Let me know if you want me to squash the commits into one, for cleaner history.

The maintainers usually require squashing.
